lib/LaTeXML/Core/Document.pm

```diff
@@ -562,15 +562,17 @@ sub serialize_attr {
 sub absorb {
   my ($self, $object, %props) = @_;
   no warnings 'recursion';
-  # Nothing? Skip it
-  my @boxes   = ($object);
+  # Record the boxes to absorb in the Document object, so that we can recover
+  # during Fatal errors by dropping the remaining ones.
+  $$STATE{boxes_to_absorb} = [$object];
+  my $boxes   = $$STATE{boxes_to_absorb};
```

Comment:
Isn't this whole `boxes_to_absorb` thing a no-op?  Although Error clears it, it's never really used.

Reply:
It is currently a no-op in the available tests yes. I would need to get extra crafty to be able to design a test for what I was trying to rescue, namely a case where a Fatal is encountered while looping over the `@$boxes` in `Document::absorb` and the flow is returned back to that loop after the Fatal is handled.

If those `$boxes` don't get reset to `[]` the loop would continue, rather than terminate right away. But none of my tests have this type of behavior captured yet.

Reply:
Nothing about tests, it's just a no-op as it is: assigning $$STATE{boxes_to_absorb}=[]` doesn't affect the local array reference `$boxes`. It would if you modified the referenced array itself, it would, but in that direction dragons lie.

We're back to the observation that we really do not want to try to blindly resume. And in the case of `absorb`, it's relatively easy: Fatal should throw (die) to the enclosing catch (eval), and you get whatever document was built up to that point.  Which, I'm guessing, is what you're trying to achieve by clearing the pending boxes (except that you don't).

Reply:
Ah, you're right, my code swaps the reference, rather than empty the existing reference. I should've written:

```perl
my $boxes_to_absorb = $$STATE{boxes_to_absorb};
@$boxes_to_absorb=() if ref $boxes_to_absorb eq 'ARRAY';
```

"Blindly resuming" has the potential to recover with less breakage than the hard die, which is why I've attempted it as a first stab. Of course it also needs to be tested well, and I'm missing a test here.

The problem with switching to `die` everywhere is that I'd need to refactor the Core.pm methods to also use `eval {}` so that e.g. the math parsing can take place after a fatal in say `absorb`. Currently a die in absorb will float all the way up to LaTeXML.pm, concluding the `Core::convertDocument` phase right then and there. The returns were trying to dance around that.

Reply:
> in that direction dragons lie.

I was also hoping I could enumerate and tame these particular dragons. If a soft `return` from a fatal ends up creating/stepping on a trap door, I can currently imagine three scenarios:
1. processing hits a harsher issue, triggering an unrecoverable fatal (such as a perl die), and ends up `die()`-ing up to the LaTeXML.pm eval anyway, a little bit later
2. or, a new class of infinite loops gets introduced in situations I haven't yet thought about. Then latexml either
   -  reaches a predefined timeout and ends up with a hard `die` from the alarm or
   - the user gets fed up and interrupts the process, which also should trigger a hard `die`.

Maybe that's all too clunky / too much boilerplate for anyone else to be convinced in. Just explaining where I was coming from.